crates/ruma-state-res/src/event_auth.rs

```diff
@@ -428,186 +362,176 @@ fn check_room_power_levels(
+        let new_power_level_too_big =
+            new_power_level.unwrap_or_else(|| field.default_value()) > sender_power_level;
+
+        if current_power_level_too_big || new_power_level_too_big {
+            return Err(format!(
+                "sender does not have enough power to changed the `{field}` power level"
+            ));
         }
     }
 
-    // FIXME: this only checks the value of the room key, it should check all values in the
-    // notifications object.
+    // Since v1, for each entry being changed in, or removed from, the events property:
+    // - Since v1, if the current value is higher than the sender's current power level, reject.
+    // Since v1, for each entry being added to, or changed in, the events property:
+    // - Since v1, if the new value is higher than the sender's current power level, reject.
+    let current_events = current_room_power_levels_event.events(room_version)?;
+    check_power_level_maps(
+        current_events.as_ref(),
+        new_events.as_ref(),
+        &sender_power_level,
+        None,
+        |ev_type| {
+            format!(
+            "sender does not have enough power to change the `{ev_type}` event type power level"
+        )
+        },
+    )?;
+
+    // Since v6, for each entry being changed in, or removed from, the notifications
+    // property:
+    // - Since v6, if the current value is higher than the sender's current power level, reject.
+    // Since v6, for each entry being added to, or changed in, the notifications property:
+    // - Since v6, if the new value is higher than the sender's current power level, reject.
     if room_version.limit_notifications_power_levels {
-        let current_value = current_room_power_levels_content.notifications.room;
-        let new_value = new_room_power_levels_content.notifications.room;
-        if current_value != new_value {
-            // Since v6, for each entry being changed in, or removed from, the notifications
-            // property:
-            // - Since v6, if the current value is higher than the sender's current power level,
-            //   reject.
-            let current_value_too_big = current_value > sender_level;
-            // Since v6, for each entry being added to, or changed in, the notifications property:
-            // - Since v6, if the new value is higher than the sender's current power level, reject.
-            let new_value_too_big = new_value > sender_level;
-            if current_value_too_big || new_value_too_big {
-                warn!("m.room.power_level failed to add ops > than own");
-                return Ok(false); // cannot add ops greater than own
-            }
-        }
+        let current_notifications = current_room_power_levels_event.notifications(room_version)?;
+        check_power_level_maps(
+            current_notifications.as_ref(),
+            new_notifications.as_ref(),
+            &sender_power_level,
+            None::<&str>,
```

Comment:
I don't think this is necessary.
```suggestion
            None,
```

Reply:
But it is.

Reply:
It doesn't matter anymore, I changed the code to fix the next comment.